sdk/dotnet/Pulumi.Automation.Tests/LocalWorkspaceTests.cs

```diff
@@ -1554,6 +1554,80 @@ public async Task InlineProgramLoggerCanRedirectToTestOutput()
             await stack.DestroyAsync();
         }
 
+        [Fact]
+        public async Task InlineProgramExceptionDuringUpShouldNotDeleteResources()
+        {
+            var program = PulumiFn.Create(() =>
+            {
+                var config = new Config();
+                var a = new ComponentResource("test:res:a", "a", null);
+
+                if (config.GetBoolean("ShouldFail") == true)
+                    throw new FileNotFoundException("ShouldFail");
+
+                var b = new ComponentResource("test:res:b", "b", null);
+                var c = new ComponentResource("test:res:c", "c", null);
+            });
+            Assert.IsType<PulumiFnInline>(program);
+
+            var stackName = $"{RandomStackName()}";
+            var projectName = "test_optionally_failing_stack";
+            using var stack = await LocalWorkspace.CreateStackAsync(new InlineProgramArgs(projectName, stackName, program)
+            {
+                EnvironmentVariables = new Dictionary<string, string?>()
+                {
+                    ["PULUMI_CONFIG_PASSPHRASE"] = "test",
+                }
+            });
+
+            var config = new Dictionary<string, ConfigValue>()
+            {
+                ["ShouldFail"] = new ConfigValue("false"),
+            };
+            try
+            {
+                await stack.SetAllConfigAsync(config);
+
+                // pulumi up
+                var upResult = await stack.UpAsync();
+                Assert.Equal(UpdateKind.Update, upResult.Summary.Kind);
+                Assert.Equal(UpdateState.Succeeded, upResult.Summary.Result);
+                Assert.True(upResult.Summary.ResourceChanges!.TryGetValue(OperationType.Create, out var upCount));
+                Assert.Equal(4, upCount);
+
+                config["ShouldFail"] = new ConfigValue("true");
+
+                await stack.SetAllConfigAsync(config);
+                var sb = new System.Text.StringBuilder();
+
+                await Assert.ThrowsAsync<FileNotFoundException>(
+                    () => stack.UpAsync(new UpOptions()
+                    {
+                        OnStandardOutput = msg => sb.AppendLine(msg),
+                    }));
+
+                var upOutput = sb.ToString();
+                Assert.DoesNotContain("test:res:b b deleted", upOutput);
```

Comment:
This is a little creative I like it :) Is there an API to do this more naturally like ResourceChanges or there is no such API?

Reply:
This test was borrowed from @vipentti's PR! I agree that it was creative. It works for our purpose here so I didn't look into improving on it. I guess we could subscribe to engine events and verify that no destroy events were published for the relevant resources?

Reply:
I'm not worried the test is fragile, I'd be worried if the API is lacking to get this info in a typed fashion if you want to. I guess there's an API so we're good.